### PR TITLE
IGNITE-20039 Add functionality to switch the IndexManager to catalog events #2

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-20039

### CODE REVIEW

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/Catalog.java [hunk 1]
```diff
@@ -115,14 +116,22 @@ public Collection<CatalogSchemaDescriptor> schemas() {
         return schemasByName.values();
     }
 
-    public CatalogTableDescriptor table(int tableId) {
+    public @Nullable CatalogTableDescriptor table(int tableId) {
```
Comment (sashapolo) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/Catalog.java:119]: Why are there no javadocs in this class?

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java [hunk 2]
```diff
@@ -254,6 +270,11 @@ public int activeCatalogVersion(long timestamp) {
         return catalogAt(timestamp).version();
     }
 
+    @Override
+    public int latestCatalogVersion() {
+        return catalogByVer.lastEntry().getKey();
```
Comment (sashapolo) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java:275]: What of `catalogByVer` map is empty?

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogService.java [hunk 3]
```diff
@@ -40,15 +41,21 @@ public interface CatalogService {
 
     String DEFAULT_ZONE_NAME = "Default";
 
-    CatalogTableDescriptor table(String tableName, long timestamp);
+    @Nullable CatalogTableDescriptor table(String tableName, long timestamp);
```
Comment (sashapolo) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogService.java:44]: Same question about javadocs

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/storage/UpdateLogImplTest.java [hunk 4]
```diff
@@ -56,69 +60,79 @@ class UpdateLogImplTest {
...
         metastore.start();
     }
 
     @AfterEach
     public void tearDown() throws Exception {
-        metastore.stop();
-        vault.stop();
+        IgniteUtils.closeAll(
+                metastore == null ? null : metastore::stop,
+                vault == null ? null : metastore::stop
```
Comment (sashapolo) [inline modules/catalog/src/test/java/org/apache/ignite/internal/catalog/storage/UpdateLogImplTest.java:75]:
```suggestion
                vault == null ? null : vault::stop
```

Comment (tkalkirill) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/storage/UpdateLogImplTest.java:75]: fix it

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java (hunk 2 above)
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java:275]: At the `CatalogManagerImpl#start`, we add a "zero version" and also check this in the `CatalogManagerSelfTest#testLatestCatalogVersion`.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/Catalog.java (hunk 1 above)
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/Catalog.java:119]: I'm not the author of this class, I don't know why there is no documentation, can I create a ticket to add documentation or should I add documentation for my methods or for the entire class?

**Review by sashapolo — APPROVED** [review]

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/Catalog.java (hunk 1 above)
Comment (sashapolo) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/Catalog.java:119]: Let's do it in a separate ticket

Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/Catalog.java:119]: Created ticket https://issues.apache.org/jira/browse/IGNITE-20046

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java [hunk 5]
```diff
@@ -1114,7 +1121,7 @@ public void operationWillBeRetriedFiniteAmountOfTimes() {
 
         doNothing().when(updateLogMock).registerUpdateHandler(updateHandlerCapture.capture());
 
-        CatalogServiceImpl service = new CatalogServiceImpl(updateLogMock, clockWaiter);
+        CatalogManagerImpl service = new CatalogManagerImpl(updateLogMock, clockWaiter);
```
Comment (AMashenkov) [inline modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java:1124]:
```suggestion
        CatalogManagerImpl manager = new CatalogManagerImpl(updateLogMock, clockWaiter);
```

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/Catalog.java [hunk 6]
```diff
@@ -115,14 +116,22 @@ public Collection<CatalogSchemaDescriptor> schemas() {
...
-    public CatalogIndexDescriptor index(int indexId) {
+    public Collection<CatalogTableDescriptor> tables() {
+        return tablesById.values();
+    }
+
+    public @Nullable CatalogIndexDescriptor index(int indexId) {
         return indexesById.get(indexId);
     }
 
+    public Collection<CatalogIndexDescriptor> indexes() {
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/Catalog.java:131]: What purpose do we need this for?
Why there is no method `indexes(int tableId)` ?

Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/Catalog.java:131]: > What purpose do we need this for?
I need to create storage for indexes at the start of the `IndexManager`.

> Why there is no method indexes(int tableId) ?
At the moment I don't need it.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogService.java [hunk 7]
```diff
@@ -40,15 +41,21 @@ public interface CatalogService {
...
-    CatalogTableDescriptor table(int tableId, int catalogVersion);
+    @Nullable CatalogTableDescriptor table(int tableId, int catalogVersion);
 
-    CatalogIndexDescriptor index(String indexName, long timestamp);
+    Collection<CatalogTableDescriptor> tables(int catalogVersion);
 
-    CatalogIndexDescriptor index(int indexId, long timestamp);
+    @Nullable CatalogIndexDescriptor index(String indexName, long timestamp);
+
+    @Nullable CatalogIndexDescriptor index(int indexId, long timestamp);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogService.java:54]: I hope that they will leave, but this is not the ticket to get rid of them.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogService.java [hunk 8]
```diff
@@ -64,5 +71,7 @@ public interface CatalogService {
 
     int activeCatalogVersion(long timestamp);
 
+    int latestCatalogVersion();
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogService.java:74]: How do you propose to change?
It turns out this `org.apache.ignite.internal.catalog.CatalogManagerImpl#activeCatalogVersion` is also not correct?

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java (hunk 5 above)
Comment (tkalkirill) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java:1124]: fix it

**Review by AMashenkov — APPROVED** [review]

**Pull request merged**